_posts/2024-04-26-modules-101.markdown

```diff
@@ -0,0 +1,135 @@
+}
+
+int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
+    VALKEYMODULE_NOT_USED(argv);
+    VALKEYMODULE_NOT_USED(argc);
+    if (ValkeyModule_Init(ctx,"module1",1,VALKEYMODULE_APIVER_1) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    if (ValkeyModule_CreateCommand(ctx,"module1.hello", hello,"",0,0,0) 
+        == VALKEYMODULE_ERR) return VALKEYMODULE_ERR;
+    return VALKEYMODULE_OK;
+}
+```
+
+All we are doing is initializing `module1` and creating a command `hello` that returns `world1`.  Now we need to update `src/modules/Makefile`
+
+```makefile
+all: ... module1.so
+
+module1.xo: ../valkeymodule.h
+
+module1.so: module1.xo
+	$(LD) -o $@ $^ $(SHOBJ_LDFLAGS) $(LIBS) -lc
+```
+
+Run `make` inside `src/modules` folder.  We will see numerous `*.so` and `*.xo` files created in `src/modules` including `module1.so`.  
+
+
+## Hello World module in Rust
+
+We will create a new Rust package by running `cargo new --lib module2` in bash.  Inside the `module2` folder we will have `Cargo.toml` and `src/lib.rs` files.  To install the valkey-module SDK run `cargo add valkey-module`.  It will create or update the `Cargo.lock` file.  
+
+Modify `Cargo.toml` to specify crate-type.  Read [Rust docs](https://doc.rust-lang.org/reference/linkage.html) to understand more about `crate-type`.
+
+```
+[lib]
+crate-type = ["cdylib"]
+```
+
+Now in `src/lib.rs` replace the existing code with the following:
+
+```rust
+#[macro_use]
+extern crate valkey_module;
+
+use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};
+
+fn hello(_ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
+    Ok(ValkeyValue::SimpleStringStatic("world2"))
+}
+
+valkey_module! {
+    name: "module2",
+    version: 1,
+    allocator: (valkey_module::alloc::ValkeyAlloc, valkey_module::alloc::ValkeyAlloc),
+    data_types: [],
+    commands: [
+        ["module2.hello", hello, "", 0, 0, 0],
+    ]
+}
+```
+
+Rust syntax is a bit different than C but we are creating `module2` with command `hello` that returns `world2` string.  We are using external crate `valkey_module` with [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and passing it variables like `name` and `version`.  Some variables like `data_types` and `commands` are arrays and we can pass zero, one or many values.  Since we are not using ctx or args we prefix them with `_` (Rust convention) instead of using `VALKEYMODULE_NOT_USED` as we did in C.  
+
+Run `cargo build` in the root folder.  We will now see `target/debug/libmodule2.dylib` (on macOS).  Build will produce *.so files on Linux and *.dll files on Windows.  
+
+
+## Run Valkey server with both modules
+
+Go back into the Valkey repo folder and run make to compile the Valkey code.  Then add these lines to the bottom of the `valkey.conf` file.
+
+```
+loadmodule path_to_valkey/src/modules/module1.so
+loadmodule path_to_module2/target/debug/libmodule2.dylib
```

Comment:
Should probably clarify the "path_to_*" needs to be updated.

Reply:
updated